src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ace/AceEditorIdleCommands.java

```diff
@@ -154,6 +168,250 @@ private void onPreviewImage(AceEditor editor,
+         
+         @Override
+         public void onRenderFinished(RenderFinishedEvent event)
+         {
+            int width = editor.getWidget().getOffsetWidth();
+            if (width == width_)
+               return;
+            
+            width_ = width;
+            renderTimer.schedule(50);
+         }
+      }));
+      
+      // initialize doc changed handler
+      docChangedHandler.set(editor.addDocumentChangedHandler(new DocumentChangedEvent.Handler()
+      {
+         String href_;
+         final Timer refreshImageTimer = new Timer()
+         {
+            @Override
+            public void run()
+            {
+               // if the discovered href isn't an image link, just bail
+               if (!isImageHref(href_))
+                  return;
+               
+               // set new src location (load handler will replace label as needed)
+               container.setWidget(new SimplePanel());
+               noImageLabel.setText("(No image at path " + href_ + ")");
+               image.getElement().setAttribute("src", imgSrcPathFromHref(href_));
+            }
+         };
+         
+         @Override
+         public void onDocumentChanged(DocumentChangedEvent event)
+         {
+            int row = plw.get().getRow();
+            Range range = event.getEvent().getRange();
+            if (range.getStart().getRow() <= row && row <= range.getEnd().getRow())
+            {
+               String line = editor.getLine(row);
+               if (isStandaloneMarkdownLink(line))
+               {
+                  // check to see if the URL text has been updated
+                  Token hrefToken = null;
+                  JsArray<Token> tokens = editor.getSession().getTokens(row);
+                  for (Token token : JsUtil.asIterable(tokens))
+                  {
+                     if (token.hasType("href"))
+                     {
+                        hrefToken = token;
+                        break;
+                     }
+                  }
+                  
+                  if (hrefToken == null)
+                     return;
+                  
+                  href_ = hrefToken.getValue();
+                  refreshImageTimer.schedule(700);
+               }
+               else
+               {
+                  onDetach.execute();
+               }
+            }
+         }
+      }));
+      
+      ChunkOutputHost host = new ChunkOutputHost()
+      {
+         @Override
+         public void onOutputRemoved(final ChunkOutputWidget widget)
+         {
+            onDetach.execute();
+         }
+         
+         @Override
+         public void onOutputHeightChanged(ChunkOutputWidget widget,
+                                           int height,
+                                           boolean ensureVisible)
+         {
+            onResize.execute(height);
+         }
+      };
+      
+      cow.set(new ChunkOutputWidget(
+            StringUtil.makeRandomId(8),
```

Comment:
We should probably just pass the document id -- main issue is that we don't have a mechanism for looking up the EditingTarget associated with an AceEditor instance (if any), and hence no way of accessing the document id.